app/src/main/java/com/toandv/demo/musicplayer/services/MusicPlayer.kt

```diff
@@ -0,0 +1,146 @@
+package com.toandv.demo.musicplayer.services
+
+import android.content.Context
+import android.media.AudioAttributes
+import android.media.MediaPlayer
+import android.net.Uri
+import android.os.PowerManager
+import com.toandv.demo.musicplayer.data.Song
+import java.util.*
+
+class MusicPlayer(private val context: Context, private val listener: PlayerListener) {
+
+    private val player: MediaPlayer = MediaPlayer()
+    private val playList = mutableListOf<Song>()
+    private lateinit var timer: Timer
+    private var isPrepared = false
+    private var pos = -1
+
+    init {
+        player.setWakeMode(context, PowerManager.PARTIAL_WAKE_LOCK)
+        player.setAudioAttributes(
+            AudioAttributes
+                .Builder()
+                .setUsage(AudioAttributes.USAGE_MEDIA)
+                .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC).build()
+        )
+        player.run {
+            setOnPreparedListener {
+                isPrepared = true
+                listener.onSongPrepared(playList[pos], player.duration)
+                it.start()
+                listener.onPlayingChanged(true)
+                startUpdate()
+            }
+
+            setOnCompletionListener {
+                isPrepared = false
+                listener.onPlayingChanged(false)
+                if (!it.isLooping) next()
+                stopUpdate()
+            }
+
+            setOnErrorListener { _, what, extra ->
+                listener.onError("Error($what, $extra)")
+                stopUpdate()
+                false
+            }
+        }
+    }
+
+    fun updatePlayList(list: List<Song>) {
+        playList.clear()
+        playList.addAll(list)
+        if (pos == -1 && playList.isNotEmpty()) pos = 0
+    }
+
+    fun setSong(id: Long): Boolean {
+        pos = playList.indexOfFirst {
+            it.id == id
+        }
+        return when (pos) {
+            -1 -> false
+            else -> {
+                playCurrentSong()
+                true
+            }
+        }
+    }
+
+    fun playOrPause() {
+        when {
+            player.isPlaying -> {
+                player.pause()
+                listener.onPlayingChanged(false)
+                stopUpdate()
+            }
+            isReady && player.isPlaying.not() -> when {
+                isPrepared -> {
+                    player.start()
+                    listener.onPlayingChanged(true)
+                    startUpdate()
+                }
+                else -> playCurrentSong()
+            }
+        }
+
+    }
+
+    fun loop() {
+        player.isLooping = !player.isLooping
+    }
+
+    fun next() {
+        when {
+            isReady.not() -> return
+            pos < playList.lastIndex -> pos++
+            else -> pos = 0
+        }
+        playCurrentSong()
+    }
+
+    fun previous() {
+        when {
+            isReady.not() -> return
+            pos == 0 -> pos = playList.lastIndex
+            else -> pos--
+        }
```

Comment:
Đơn giản thì dùng if thôi thay vì dùng when
```suggestion
        if(isReady.not()) return
        val size = playList.size
        pos = (pos + size -1) % size
```